sycl/include/sycl/detail/pi.h

```diff
@@ -1130,11 +1130,13 @@ __SYCL_EXPORT pi_result piextQueueCreateWithNativeHandle(
 //
 // Memory
 //
-__SYCL_EXPORT pi_result piMemBufferCreate(
-    pi_context context, pi_mem_flags flags, size_t size, void *host_ptr,
-    pi_mem *ret_mem, const pi_mem_properties *properties = nullptr);
+__SYCL_EXPORT pi_result
+piMemBufferCreate(pi_context context, pi_device device, pi_mem_flags flags,
```

Comment:
This is the device, the memory will be allocated on. Even if we are putting multiple CUDA devices in a single context, they still have distinct device memories, so for the allocation we need to know what device to allocate on. It needs to be given and the runtime is changed in this PR so that it is always given. On backends that do not need this information device argument can be ignored.

Reply:
The buffer would be represented by multiple allocations, one on each device in the context. It would then "migrate" (copy between allocations) depending on where the buffer is being accessed from. So what is that single "device" that you are adding and why it was not necessary for OpenCL and Level-Zero but is needed for CUDA?

Reply:
No, both OpenCL and Level-Zero backends support context with multiple devices in it even if they have their distinct memories. That's why I said there multiple allocations representing such a buffer in the context, and it is being copied essentially from device where it was used the last to where it is needed next.

Reply:
I think we cannot practically "undo" OpenCL RT buffer migration, so there would be an excessive "copying" in that backend if we also do the migration in SYCL RT. Anytime a buffer is used from different devices (of the same context) there will be a copy or other way to access the memory actual on another device, e.g. P2P access if that's available. Knowing these details seem to belong to plugins, and not SYCL RT (unless we create additional more explicit interfaces).

Reply:
I agree that it is a possible way to go, but I am not sure it justifies the "complexity" of breaking backwards compatibility (btw, I am still not sure why you want to have some "initial" device for a buffer allocation since that can immediately change as app starts using it from other devices). The buffer "migration" across devices in the context implemented in OpenCL and Level-Zero is not a big deal, in my view, so how do we justify a re-course? 

Adding few folks for extra opinions: @steffenlarsen , @romanovvlad

Reply:
> There is no preemptive allocation on some device with this approach. You only have a device allocation when you are first using a buffer on that device. It can only migrate to another device after that allocation is first used on the device it was allocated on.

So this is the device that first used that buffer, and nothing more? Eventually it will be migrated (by SYCL RT in your proposal) to whatever other device that needs it in this context, right?

Reply:
Exactly. Except if the backend returns `PLUGIN_MANAGED_OR_SAME` from `piGetMemoryConnection`. In this case RT will do nothing and assume the backend can handle any migrations it needs. That can be used for any pair of devices in the same context for the OpenCL backend.

Reply:
I added the query.